### PR TITLE
Migrate to flat config

### PR DESCRIPTION
**Pull request by fisker**

### CODE REVIEW

**Comment by fisker** [comment]
`eslint-config-xo-typescript` may need expose an array instead of an object https://github.com/xojs/eslint-config-xo-typescript/blob/3271346cfa8cdc2d44bee64718399eea9f603ea3/index.js#L737

Should we expose an array here too?


So user can

```js
export {default} from 'eslint-config-xo';
```

and

```diff
import eslintConfigXo from 'eslint-config-xo';

export default [
-	eslintConfigXo,
+	...eslintConfigXo,
];
```

**Comment by fregante** [comment]
> eslint-config-xo-typescript

Do you need separate packages anymore? My understanding is that configurations are now _configurable_, so you can just expose a function and pass `typescript:true`, `spaces:4`, etc if needed.  Is that correct?

**Comment by fisker** [comment]
But eslint-config-xo-typescript need extra packages installed/imported, it's not good if user only uses js syntax.

**Comment by fisker** [comment]
> Should we expose an array here too?

Looks like shared config can be an object, maybe arrays will be flattened too, if it's true, we don't need care, I'll test.

**Comment by fregante** [comment]
> But eslint-config-xo-typescript need extra packages installed/imported, it's not good if user only uses js syntax.

That hasn't been a limiting factor for XO itself, which includes TypeScript and Prettier whether you use them or not.

**Comment by fisker** [comment]
Waiting for answers https://github.com/eslint/eslint/issues/18563

**Comment by sindresorhus** [comment]
I don't want to merge the TS one into this one yet.

The "space" config should be merged into this one though. It was only a separate package because of ESLint limitations.

**Comment by fisker** [comment]
Let's export an array?  Check discussion in https://github.com/eslint/eslint/issues/18563

**Comment by sindresorhus** [comment]
👍

**Review by voxpelli — COMMENTED** [review]
I would go with generating the config through a function instead of through static objects (see review comment) as one can now be sure that the config will be used from a javascript file and hence it can be better to expose a typed factory function rather than relying on setting up files for all the possible combinations of options.

I would also consider eventually upstreaming `eslint-config-xo` into `xo` and let it be run through either `xo` or be extended and run through `eslint`.

The `eslint-config-` prefix is no longer needed to signify that an ESLint config can be returned.

File: space-browser.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import eslintConfigXoBrowser from './browser.js';
+
+const [config] = eslintConfigXoBrowser;
+
+export default [
+	{
+		...config,
+		rules: {
+			...config.rules,
+			indent: [
```
[the remark this answers is not in the document]
Comment (voxpelli) [reply space-browser.js:10]: A benefit of having them as separate (+ including names for each one of them) is that its very clear which config that leads to which config. See example here from the config inspector generated for neostandard: https://neostandard.github.io/neostandard/configs

**Comment by fisker** [comment]
Sorry that I didn't get time to check feedbacks.

**Comment by Primajin** [comment]
Hey, do you think it makes sense to add a legacy support like they did over here in vitest: https://github.com/vitest-dev/eslint-plugin-vitest/pull/434/files

image

**Comment by voxpelli** [comment]
@Primajin Open a new issue for that discussion. Personally I would say: No, it does not make sense :)

**Pull request merged**